test/flows/next_steps_for_your_business_flow_test.rb

```diff
@@ -0,0 +1,184 @@
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of financial_support for any response" do
+        assert_next_node :financial_support, for_response: "import_goods"
+      end
+    end
+  end
+
+  context "question: financial_support" do
+    setup do
+      testing_node :financial_support
+      add_responses annual_turnover_over_85k: "yes",
+                    employ_someone: "yes",
+                    activities: "import_goods"
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of business_premises for any response" do
+        assert_next_node :business_premises, for_response: "yes"
+      end
+    end
+  end
+
+  context "question: business_premises" do
+    setup do
+      testing_node :business_premises
+      add_responses annual_turnover_over_85k: "yes",
+                    employ_someone: "yes",
+                    activities: "import_goods",
+                    financial_support: "yes"
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of results for any response" do
+        assert_next_node :results, for_response: "home"
+      end
+    end
+  end
+
+  context "outcome: results" do
+    setup do
+      testing_node :results
+      add_responses annual_turnover_over_85k: "no",
+                    employ_someone: "no",
+                    activities: "none",
+                    financial_support: "no",
+                    business_premises: "none"
+    end
+
+    should "render the default results" do
+      assert_rendered_outcome text: "Check if you need any licences for your business"
+      assert_rendered_outcome text: "Check if you need insurance"
+      assert_rendered_outcome text: "Check if you need to send a Self Assessment"
+      assert_rendered_outcome text: "Selling goods or services"
+      assert_rendered_outcome text: "Avoid anti-competitive activity"
+      assert_rendered_outcome text: "Learn how to deal with personal information"
+      assert_rendered_outcome text: "Health and safety at work"
+      assert_rendered_outcome text: "How to run a limited company"
+      assert_rendered_outcome text: "Starting a business overview"
+      assert_rendered_outcome text: "Include your company details on stationery"
+      assert_rendered_outcome text: "How to dispose of your company's waste"
+      assert_rendered_outcome text: "Protect your brand or designs"
+      assert_rendered_outcome text: "Get help from the Business Support Helpline"
+      assert_rendered_outcome text: "Sign up for Companies House email reminders"
+    end
+
+    should "render the results if not registered for corporation tax" do
+      add_responses ct: "true"
+      assert_rendered_outcome text: "Register for Corporation Tax"
```

Comment:
I don't understand how this test is passing.

It looks like `registered_for_corp_tax ` is being set to "true" if `ct` is "true" and "false" for any other value:

```ruby
calculator.registered_for_corp_tax = (ct == "true")
```

However, in the calculator, the text for `r1` should be displayed if `registered_for_corp_tax` is not true:

```ruby
r1: ->(calculator) { calculator.registered_for_corp_tax != true },
```

So I don't understand how the "Register for Corporation Tax" text is being rendered. Nor do I understand how the value for `ct` is being set in the flow.

Reply:
I've tried this manually, and it looks like "Register for Corporation Tax" does NOT display if `ct=true` is [appended to the querystring](https://www.gov.uk/next-steps-for-your-business/results?activities%5B%5D=export_go[…]B%5D=home&employ_someone=not_sure&financial_support=yes&ct=true ) but DOES display if it's [not appended](https://www.gov.uk/next-steps-for-your-business/results?activities%5B%5D=export_go[…]B%5D=home&employ_someone=not_sure&financial_support=yes) so I'm still confused.

Reply:
I changed the response in the test to `add_responses ct: "false"` and it still passes, so I think we need to re-think how the value if `ct` is set in the test. 

Perhaps @kevindew has some ideas?

Reply:
Ah sooo this is super confusing, because I made a mistake - the test should have `add_response ct: "false"`.

But it's weird that it's still passing - I'll take a look.

---
Just an explainer for the behaviour:
```
calculator.registered_for_corp_tax = (ct == "true")
```

is setting the the attribute `registered_for_corp_tax` to a boolean if the ct query parameters is a String "true". (i.e. type casting)

Then the rule to show the result:
```
r1: ->(calculator) { calculator.registered_for_corp_tax != true }
```

returns false, if the user has already registered for corp tax. (e.g. ct="true"). This is because the result is "Register for corporation tax", which we only want to show user's that haven't already. Overall the behaviour is hide the results for users with `ct=true` appended to the query parameters.

Btw, the context is Companies House sends links to our tool to users. They generate a link with the `ct` query parameter already set.

Reply:
I was just looking at this after @leenagupte's tag so I can save you a little time @theseanything.

The lack of this doing anything is caused by string/symbol key indexing of the underlying hash in response store. 
Basically: 
https://github.com/alphagov/smart-answers/blob/48c9b5d5a507dbc5fcb66cc9133a14333ce25b28/lib/smart_answer/state_resolver.rb#L9-L11
looks up the response store by a symbol

whereas:
https://github.com/alphagov/smart-answers/blob/48c9b5d5a507dbc5fcb66cc9133a14333ce25b28/lib/smart_answer/state_resolver.rb#L38-L40

uses a string.

This only affects the tests because the data that goes into the [response store in tests is a regular hash](https://github.com/alphagov/smart-answers/blob/48c9b5d5a507dbc5fcb66cc9133a14333ce25b28/test/support/flow_test_helper.rb#L85), whereas in [FlowHelper](https://github.com/alphagov/smart-answers/blob/main/app/helpers/flow_helper.rb#L16) it's a HashWithIndifferentAccess.

So we can sort this out if we ensure that a hash with indifferent access is used consistently in the response store class. 

There'd be a few ways to resolve that, I wonder if one of the best ones is to call `.with_indifferent_access` on the initial argument in ResponseStore as I believe that will also dup the hash (which seems like desirable behaviour as you'd not want anything to be actually modifying request.query_parameters which is what is probably already happening)

Reply:
Thanks for the explanation @theseanything. So it looks like the default behaviour is to display the "Register for corporation tax" text, so this check could probably be added to the first test. It looks like what we need for this test is to check that the text _isn't_ being displayed.

Reply:
Thanks both! I agree I think adding `.with_indifferent_access` might be the thing to do - I'll add a commit to fix that here.

Happy to move this into the first test - however, I'm worried about adding tests to show that results are hidden. If I do this for this result, then why not all the others? Which essential takes closer to trying to test every possible permutation of results - which is going to be super cumbersome for these node tests (and dups of the calculator unit tests).

Reply:
I think the test for "hidden" results is needed here, because that's the purpose of the `ct` parameter. How will we know if the `. with_indifferent_access` changes are working if don't test it?

Reply:
But you could argue that for any result - e.g. the financial support results are hidden if the user answers "no" etc?

If we want to verify the behaviour of the test helpers, they'd need to their own unit tests. These tests are for testing the flow not the test helpers.

Reply:
I agree, these tests are for the for flow, and this behaviour, which I struggled to understand is, set in the flow:

```ruby
calculator.registered_for_corp_tax = (ct == "true")
```

I don't believe there are any plans to test it elsewhere?

Reply:
Ah right, sorry I thought you meant about testing the response store for a HashWithIndifferentAccess behaviour rather than a Hash.

As for that line, I think what I'm struggling with is _what are we trying to test_ with for these node based tests? Are we unit testing the on_response, next_node etc code blocks? Are we unit testing the page renders?

That particular line gets executed in the default test where the results is displayed. However, we don't have a test specifically for the permutation where `ct="true"`. But if I were to add that in, why would I not add in a test for all the other permutation of the other results? E.g.employ_someone="in_future" shows the "Check what to do before employing someone" result.

I'm just unclear around the reasoning around what assertions we need to make.

(Also not sure how to assert that text is not present - guessing I'd have to add another test helper?)

Reply:
I can help on this. I'll drop a reply once I'm out of a meeting.

Reply:
**TL;DR - no tests for outcome would meet the standard for testing, a few outcome tests would make this a sufficiently thorough integration test**

---

Great and interesting discussion here, it often seems that deciding the boundaries of what to test varies on personal perspective. I find it's often good to think in terms of [necessary and sufficient tests] (https://blog.testdouble.com/posts/2020-02-25-necessary-and-sufficient/) to try to establish what the boundaries should be.

[As referenced in the ADR](https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md#decision) the purpose of outcome tests are to explicitly test specific logic in an outcome. Motivation for this is mostly coverage related, and it's really on the basic side of testing that a lot of Smart Answer's lack (for an example of why just this level of basic is enough, yet not currently met, is https://github.com/alphagov/smart-answers/blob/0f1daad3d148b384ce65e06ae738a759a8315a61/app/flows/maternity_paternity_calculator_flow/outcomes/adoption_not_entitled_to_leave_or_pay.erb#L5 where `a_leave_employment_threshold` will raise a variable not defined error).

There is not specific logic in the outcome, so I'd argue it's not necessary that we actually have any outcome tests. We achieve the goal of the code being executed by at least one of `calculator.grouped_results` returning true. So I'd be satisfied that having a unit test for `grouped_results` is enough that we're confident we have the necessary tests to meet the outcome testing goal.

The more ambiguous area seems to be deciding if the tests we have are sufficient. In that case we'll want to consider the domain of this individual test and the impact of individual lines of code in the flow. So just the basic testing thing of "if this line is code is removed will something break" (TBH I doubt all Smart Answers will 100% comply with this level of testing due to the tedium involved in some cases).

It could also be deciding that we want these tests to assert that the YAML file integrates with flow. My view would be that testing YAML file integration can be done in the calculator unit tests so doesn't belong here.

I'd consider having a few assertions to do with the 6 attributes that are stored on the calculator to be enough for these tests to be sufficient (example below). I think they do enough to confirm that the flow integrates with the calculator and then anything further needed should be testable at calculator level.

```
  context "outcome: results" do
    setup do
      testing_node :results
      add_responses annual_turnover_over_85k: "no",
                    employ_someone: "no",
                    activities: "none",
                    financial_support: "no",
                    business_premises: "none"
    end

    should "render guidance specific to annual turnover" do
      add_responses annual_turnover_over_85k: "yes"
      assert_rendered_outcome text: "Register for VAT"    
    end

    should "hide corporation tax guidance based on ct parameter" do
      assert_match "Register for Corporation Tax", @test_flow.outcome_text
      add_responses ct: "true"
      assert_no_match "Register for Corporation Tax", @test_flow.outcome_text
    end

    should "render guidance specific to employing someone" do
      add_responses employ_someone: "yes"
      assert_rendered_outcome text: "Check what to do before employing someone"    
    end

    should "render guidance specific to a businesses activities" do
      add_responses activities: %w[import_goods]
      assert_rendered_outcome text: "How to buy items from abroad"   
    end

    should "render guidance specific to financial support" do
      add_responses financial_support: "yes"
      assert_rendered_outcome text: "See what finance and support is available"
    end

    should "render guidance specific to business premise" do
      add_responses business_premises: %w[home]
      assert_rendered_outcome text: "Check the rules on running a business from home"
    end
  end
```

does this help?

Reply:
I've updated the tests, which hopefully covers at least the some permutation of the results being shown or hidden. Let me know that make sense.

Reply:
Great thank you! Yes they look very thorough.